mlir/include/mlir/Interfaces/MemOpInterfaces.td

```diff
@@ -0,0 +1,117 @@
+  let description = [{
+    An interface for operations that can consume memory-space cast-like
+    operations.
+  }];
+  let cppNamespace = "::mlir";
+  let methods = [
+    InterfaceMethod<[{
+        Attempt to bubble-down the incoming cast-like operands. On success
+        returns any new results, and whether the operation was modified in
+        place, otherwise it returns failure.
+        If new results are produced, these must be compatible with the original
+        operation results.
+
+        If the operation was not modified in place, then the interface
+        guarantees it is valid to erase the original operation.
+        If the operation was modified in place, then the interface must
+        guarantee no operations were created by the method, and that no further
+        IR modification is necessary.
+
+        Any implementations of this method must not erase/replace the original
+        operation, instead it is the caller responsibility to erase or replace
+        the op with the results provided by the method.
+
+        Finally, any implementations of this method have to guarantee that the
+        IR remains valid at all times.
+      }],
+      "::llvm::FailureOr<std::pair<::llvm::SmallVector<::mlir::Value>, bool>>",
+      "bubbleDownCasts",
+      (ins "::mlir::OpBuilder &":$builder)
+    >,
+  ];
+}
+
+def MemorySpaceCastOpInterface : OpInterface<"MemorySpaceCastOpInterface"> {
+  let description = [{
+    An interface for operations that perform memory-space casts. This
+    interface assumes that the cast operation is `pure`.
+
+    These operations expect to have a well-defined ptr-like operand, and
+    a well-defined target ptr-like result.
+  }];
+  let cppNamespace = "::mlir";
+  let methods = [
+    InterfaceMethod<[{
+        Returns the source ptr-like value.
+      }],
+      "::mlir::TypedValue<::mlir::PtrLikeTypeInterface>",  "getSourcePtr"
+    >,
+    InterfaceMethod<[{
+        Returns the target ptr-like value.
+      }],
+      "::mlir::TypedValue<::mlir::PtrLikeTypeInterface>", "getTargetPtr"
+    >,
+    InterfaceMethod<[{
+        Returns whether the memory space cast specified by `tgt` and `src`
+        is supported.
+      }],
+      "bool", "isValidMemorySpaceCast",
+      (ins "::mlir::PtrLikeTypeInterface":$tgt,
+           "::mlir::PtrLikeTypeInterface":$src)
+    >,
+    InterfaceMethod<[{
+        Clones the memory space cast op with the given source and target type.
+      }],
+      "::mlir::MemorySpaceCastOpInterface", "cloneMemorySpaceCastOp",
+      (ins "::mlir::OpBuilder &":$builder, "::mlir::PtrLikeTypeInterface":$tgt,
+           "::mlir::Value":$src)
+    >,
+    InterfaceMethod<[{
+        Returns whether the memory-space cast is lossless. A lossless
+        memory-space cast must not lose any information encoded in the memory
+        space. An example of such cast, is any conversion to the generic memory
+        space.
```

Comment:
... This is false

`memory_space_cast %x : #amdgpu.address_space<raw_fat_buffer> to [null]` isn't information-preserving - it throws away all the buffer metadata.

Similarly, on a CHERI, you could have casts that discard provedance.

Reply:
That's true it's a poorly chosen name, I'll change it.
What I meant to indicate is whether the cast can be bubbled down by taking the source. Originally, the function was called `isFusableCast`, but I changed it for worse.

Reply:
Is not only about whether the property holds. It also related to whether the cast considers the bubbling-down beneficial.

For example, in GPUs `ptr<3> -> ptr<0>` is beneficial to take the source, however `ptr<0> -> ptr<3>` it's not.